File: php/src/main/php/org/shypl/sna/SocialNetworkManager.php
<?php
namespace org\shypl\sna;

use InvalidArgumentException;
use org\shypl\sna\impl\MmSocialNetwork;
use org\shypl\sna\impl\OkSocialNetwork;
use org\shypl\sna\impl\VkSocialNetwork;

final class SocialNetworkManager {
	/**
	 * @var SocialNetwork[]
	 */
	private static $networks = array();

	/**
	 * @param SocialNetwork $network
	 */
	public static function registerNetwork(SocialNetwork $network) {
		self::$networks[] = $network;
	}

	/**
	 * @param int $id
	 *
	 * @return SocialNetwork
	 */
	public static function getNetworkById($id) {
		foreach (self::$networks as $network) {
			if ($network->getId() == $id) {
				return $network;
			}
		}
		throw new InvalidArgumentException('Network by id ' . $id . ' is not registered');
	}

	/**
	 * @param string $name
	 *
	 * @return SocialNetwork
	 */
	public static function getNetworkByName($name) {
		foreach (self::$networks as $network) {
			if ($network->getName() == $name) {
				return $network;
			}
		}
		throw new InvalidArgumentException('Network by name ' . $name . ' is not registered');
	}
}

SocialNetworkManager::registerNetwork(new VkSocialNetwork());
SocialNetworkManager::registerNetwork(new MmSocialNetwork());
SocialNetworkManager::registerNetwork(new OkSocialNetwork());

File: java/src/main/java/org/shypl/sna/SocialNetworkManager.java
package org.shypl.sna;

import org.shypl.sna.impl.FbSocialNetwork;
import org.shypl.sna.impl.MmSocialNetwork;
import org.shypl.sna.impl.OkSocialNetwork;
import org.shypl.sna.impl.VkSocialNetwork;

import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.NoSuchElementException;
import java.util.Set;

public final class SocialNetworkManager {
	private static Set<SocialNetwork> networks = new HashSet<>();

	static {
		registerNetwork(new VkSocialNetwork());
		registerNetwork(new MmSocialNetwork());
		registerNetwork(new OkSocialNetwork());
		registerNetwork(new FbSocialNetwork());
	}

	public static void registerNetwork(SocialNetwork network) {
		networks.add(network);
	}

	public static int countNetworks() {
		return networks.size();
	}

	public static Collection<SocialNetwork> getNetworks() {
		return Collections.unmodifiableCollection(networks);
	}

	public static SocialNetwork getNetwork(int id) {
		for (SocialNetwork network : networks) {
			if (network.getId() == id) {
				return network;
			}
		}
		throw new NoSuchElementException("Network by id " + id + " is not registered");
	}

	public static SocialNetwork getNetwork(String code) {
		for (SocialNetwork network : networks) {
			if (network.getCode().equals(code)) {
				return network;
			}
		}
		throw new NoSuchElementException("Network by code \"" + code + "\" is not registered");
	}
}
